### ISSUE TITLE
Eliminate redundant output buffer allocations in the IREE Execution Provider

### ISSUE THREAD

**Issue by Abhishek-Varma** [iree-org/onnxruntime-ep-iree#48]
## Summary

The IREE EP currently allocates an output buffer twice for every model invocation — once inside IREE and once in ORT — and copies between them. IREE supports caller-provided output buffers natively, and the EP should adopt that path so the compiled module writes results directly into ORT-owned (or user-owned) storage, with no extra alloc and no post-execution copy.

## Background

Today, for each output of every `Compute()` call, the EP performs:

```cpp
// onnxruntime/core/providers/iree/iree_ep.cc#L561-L564
auto& output = ctx.GetOutput(i, shape.data(), shape.size());   // (1) ORT-side allocation
IreeBufferViewToOrtTensor(iree_outputs[i], output);            // (2) copy from IREE -> ORT
```

Independently, the IREE invocation has already allocated its own buffer to hold the dispatch result. The end-to-end pattern is therefore:

| # | Allocation                | Owner        | Purpose                          |
|---|---------------------------|--------------|----------------------------------|
| 1 | IREE result buffer        | IREE runtime | Where the dispatch writes        |
| 2 | ORT output tensor         | ORT          | What `Compute()` returns to user |

…with a memcpy from (1) to (2) before each call returns.

For models with non-trivial outputs (LLM logits, segmentation maps, KV-cache slices) and high invocation rates, both the redundant allocation and the device-side copy are measurable.

## What IREE already provides

IREE supports caller-provided output storage as an ABI feature:

- A compiled module's exported function can declare that one of its inputs is in fact backing storage for one of its outputs, via the `iree.abi.output` ABI attribute on the storage argument.
- Inside the wrapper, IREE's `hal.tensor.alias` op binds the result tensor's storage to the caller-provided HAL buffer.
- When the alias pattern is recognized, IREE's stream allocator folds the producer dispatch to write directly into the caller-provided buffer, eliding the IREE-side temp allocation.

The net effect for a caller is: hand IREE the buffer you want results in (as an extra positional argument to the invocation, marked with `iree.abi.output`), and the compiled program writes its output bytes there directly.

## Approach

The IREE op that signals "this SSA tensor's storage aliases this user-provided HAL buffer" is `hal.tensor.alias`. I tried considering `hal.tensor.export` as an alternative, but its definition has no way to express such a binding, it only converts a tensor SSA value to a buffer view at the function boundary.

The intended IR pattern, synthesized inside the compiled module's wrapper, is:

```mlir
%a       = <compute>
%a_alias = hal.tensor.alias %a to %user_buff_a
%res     = hal.tensor.export %a_alias
```

This communicates the alias relationship to IREE's stream allocator *before* allocation decisions are made, which is what allows the IREE-side temp allocation to be elided in the first place.

The corresponding EP-side flow becomes:

1. Allocate the output tensor on the ORT side first.
2. Bind that storage as an IREE output buffer (using IREE's runtime API).
3. Invoke the IREE module.

With this in place the compiled program writes its result bytes directly into the ORT-managed buffer, and `IreeBufferViewToOrtTensor` is no longer needed on that path.

## Plan

We will roll this out in two phases, contingent on what shape information the EP has *before* invoking IREE.

Consequentially static shapes' support should be the easier case before we extend to dynamic shape cases.

I'll soon scope out the individual components we will need to make the use-case seamless to implement/integrate with EP.

Also, for dynamic case I was considering either of the two options :-
1. ONNX shape inference to compute output shape from graph metadata.
2. A separate "shape-only" function alongside the main function, and EP can call that first to learn about output shapes.

Based on the discussion with @IanWood1 - I'll go ahead with option 2 which is to have a "shape-only" function and refer to https://github.com/iree-org/iree/pull/22625 for the same. But as a first milestone I'll work towards having an e2e static shape example checked in EP.

CC: @MaheshRavishankar @IanWood1 @jtuyls
